File: solutions/2026-05-20/java/2026-05-20-find-the-prefix-common-array-of-two-arrays.java
```java
import java.util.HashSet;
import java.util.Set;

class Solution {
    /**
     * Given two permutations A and B of length n, find the prefix common array C
     * where C[i] is the count of numbers present in the first i+1 elements of both A and B.
     * Problem Link: https://leetcode.com/problems/find-the-prefix-common-array-of-two-arrays/
     *
     * Approach:
     * We can iterate through the arrays A and B simultaneously. At each index `i`,
     * we need to count how many numbers from `A[0...i]` are also present in `B[0...i]`.
     * To efficiently check for presence, we can use two HashSets. One set (`seenA`)
     * will store elements encountered in `A` up to the current index `i`, and another
     * set (`seenB`) will store elements encountered in `B` up to the current index `i`.
     *
     * For each index `i`:
     * 1. Add `A[i]` to `seenA`.
     * 2. Add `B[i]` to `seenB`.
     * 3. Initialize a `commonCount` to 0.
     * 4. Iterate through the elements in `seenA`. For each element, check if it is
     *    also present in `seenB`. If it is, increment `commonCount`.
     * 5. Store `commonCount` in the result array `C` at index `i`.
     *
     * Time Complexity: O(n^2) in the worst case. For each of the `n` iterations,
     * we iterate through `seenA` which can grow up to size `n`. So, `n * n = n^2`.
     * Space Complexity: O(n) for storing elements in the two HashSets.
     *
     * Optimization for Time Complexity:
     * The above approach recalculates the common count from scratch at each step.
     * A more optimized approach is to maintain the common count incrementally.
     *
     * Optimized Approach:
     * We can maintain two sets: `seenA` for elements in A seen so far, and `seenB` for elements in B seen so far.
     * We also maintain a `commonCount` variable.
     * For each index `i`:
     * 1. Add `A[i]` to `seenA`.
     * 2. Add `B[i]` to `seenB`.
     * 3. Check if `A[i]` is present in `seenB`. If it is, increment `commonCount`.
     * 4. Check if `B[i]` is present in `seenA`. If it is AND `A[i]` was not equal to `B[i]`, increment `commonCount`.
     *    This avoids double counting when `A[i] == B[i]`.
     * 5. Store `commonCount` in `C[i]`.
     *
     * This optimized approach leads to:
     * Time Complexity: O(n) because we iterate through the arrays once, and set operations (add, contains) are O(1) on average.
     * Space Complexity: O(n) for the two HashSets.
     */
    public int[] findThePrefixCommonArray(int[] A, int[] B) {
        int n = A.length;
        int[] C = new int[n]; // The prefix common array to store the results

        // Use HashSets for efficient lookups (average O(1) for add and contains)
        Set<Integer> seenA = new HashSet<>();
        Set<Integer> seenB = new HashSet<>();
        
        int commonCount = 0; // Variable to keep track of common elements

        // Iterate through both arrays up to index n-1
        for (int i = 0; i < n; i++) {
            // Add the current element from A to its set
            seenA.add(A[i]);
            // Add the current element from B to its set
            seenB.add(B[i]);

            // Check if the current element from A is present in the seen elements of B
            // If it is, it means this number is now common in the prefixes A[0..i] and B[0..i]
            if (seenB.contains(A[i])) {
                commonCount++;
            }
            // Check if the current element from B is present in the seen elements of A
            // If it is, and if A[i] and B[i] are not the same element (to avoid double counting
            // when A[i] == B[i] and A[i] was already added to seenB), it means this number
            // is also common in the prefixes A[0..i] and B[0..i].
            if (seenA.contains(B[i]) && A[i] != B[i]) {
                commonCount++;
            }
            
            // Store the current total count of common elements for prefix up to index i
            C[i] = commonCount;
        }

        return C; // Return the computed prefix common array
    }
}
```

File: solutions/2026-05-20/python/2026-05-20-find-the-prefix-common-array-of-two-arrays.py
```python
# Problem: Find the Prefix Common Array of Two Arrays
# Link: https://leetcode.com/problems/find-the-prefix-common-array-of-two-arrays/
# Approach:
# We need to calculate a prefix common array C where C[i] is the count of common elements
# within the first i+1 elements of arrays A and B.
# We can iterate through both arrays simultaneously from index 0 to n-1.
# At each index i, we consider the subarrays A[0...i] and B[0...i].
# To efficiently count common elements, we can use sets.
# We maintain two sets, `seen_a` and `seen_b`, to store elements encountered so far in A and B respectively.
# For each index `i`:
# 1. Add A[i] to `seen_a`.
# 2. Add B[i] to `seen_b`.
# 3. The common elements between A[0...i] and B[0...i] are the intersection of `seen_a` and `seen_b`.
# 4. The count of common elements is the size of this intersection. This count is C[i].
# Time Complexity: O(n), where n is the length of the arrays. We iterate through the arrays once. Set insertions and intersections take O(1) on average for the given constraints (n <= 50).
# Space Complexity: O(n), for storing the elements in the two sets `seen_a` and `seen_b`.
class Solution:
    def findThePrefixCommonArray(self, A: list[int], B: list[int]) -> list[int]:
        n = len(A)
        # Initialize the prefix common array C with zeros
        C = [0] * n
        # Initialize sets to keep track of elements seen so far in A and B
        seen_a = set()
        seen_b = set()

        # Iterate through the arrays from index 0 to n-1
        for i in range(n):
            # Add the current element from A to seen_a
            seen_a.add(A[i])
            # Add the current element from B to seen_b
            seen_b.add(B[i])

            # Calculate the intersection of seen_a and seen_b.
            # This gives us the common elements present in A[0...i] and B[0...i].
            common_elements = seen_a.intersection(seen_b)

            # The count of common elements is the length of the intersection set.
            # This count is stored at C[i].
            C[i] = len(common_elements)

        # Return the calculated prefix common array
        return C

```